core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/sendsemantics/partitioners/RangeBasedShufflePartitioner.scala

```diff
@@ -0,0 +1,42 @@
+package edu.uci.ics.amber.engine.architecture.sendsemantics.partitioners
+
+import edu.uci.ics.amber.engine.architecture.sendsemantics.partitionings.RangeBasedShufflePartitioning
+import edu.uci.ics.amber.engine.common.tuple.ITuple
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+import edu.uci.ics.texera.workflow.common.tuple.schema.AttributeType
+
+case class RangeBasedShufflePartitioner(partitioning: RangeBasedShufflePartitioning)
+    extends ParallelBatchingPartitioner(partitioning.batchSize, partitioning.receivers) {
+
+  override def selectBatchingIndex(tuple: ITuple): Int = {
+    // Do range partitioning only on the first attribute in `rangeColumnIndices`.
+    val fieldType = tuple
+      .asInstanceOf[Tuple]
+      .getSchema
+      .getAttributes()
+      .get(partitioning.rangeColumnIndices(0))
+      .getType
+    var fieldVal: Long = -1
+    fieldType match {
+      case AttributeType.LONG =>
+        fieldVal = tuple.getLong(partitioning.rangeColumnIndices(0))
+      case AttributeType.INTEGER =>
+        fieldVal = tuple.getInt(partitioning.rangeColumnIndices(0)).toLong
+      case AttributeType.DOUBLE =>
+        fieldVal = tuple.getDouble(partitioning.rangeColumnIndices(0)).toLong
+      case _ =>
+        throw new RuntimeException("unsupported attribute type: " + fieldType.toString())
+    }
+
+    if (fieldVal < partitioning.rangeMin) { return 0 }
+    if (fieldVal > partitioning.rangeMax) { return partitioning.receivers.length - 1 }
+    var destinationReceiver: Int = 0
+    val keysPerReceiver =
+      ((partitioning.rangeMax - partitioning.rangeMin) / partitioning.receivers.length).toLong + 1
+    while (fieldVal >= partitioning.rangeMin + keysPerReceiver * destinationReceiver) {
```

Comment:
Can `destinationReceiver` be directly calculated using `floor(fieldVal/keysPerReceiver)`?

Reply:
No. Eg: rangeMin = -900, rangeMax = -100, keysPerReceiver = 100, and fieldVal = -753. Then destinationReceiver should be  1, not -753/100. Essentially, rangeMin has a role to play.

Reply:
but I can do something using floor to simplify the calculation, if that's what you mean

Reply:
done

Reply:
yes, that's what I mean